go-runtime/modulecontext/module_context.go

```diff
@@ -2,59 +2,206 @@ package modulecontext
+type contextKeyModuleContext struct{}
+
+func New() *ModuleContext {
 	return &ModuleContext{
-		module:         module,
-		configManager:  cm,
-		secretsManager: sm,
-		dbProvider:     dbp,
+		configs:   map[string][]byte{},
+		secrets:   map[string][]byte{},
+		databases: map[string]dbEntry{},
+		mockVerbs: map[Ref]MockVerb{},
+	}
+}
+
+func NewForTesting() *ModuleContext {
+	moduleCtx := New()
+	moduleCtx.isTesting = true
+	return moduleCtx
+}
+
+func FromContext(ctx context.Context) *ModuleContext {
+	m, ok := ctx.Value(contextKeyModuleContext{}).(*ModuleContext)
+	if !ok {
+		panic("no ModuleContext in context")
 	}
+	return m
 }
 
-// ApplyToContext returns a Go context.Context that includes configurations,
-// secrets and DSNs can be retreived Each of these components have accessors to
-// get a manager back from the context
+// ApplyToContext returns a Go context.Context with ModuleContext added.
 func (m *ModuleContext) ApplyToContext(ctx context.Context) context.Context {
-	ctx = ContextWithDBProvider(ctx, m.dbProvider)
-	ctx = cf.ContextWithConfig(ctx, m.configManager)
-	ctx = cf.ContextWithSecrets(ctx, m.secretsManager)
-	return ctx
+	return context.WithValue(ctx, contextKeyModuleContext{}, m)
 }
 
-// FromEnvironment creates a ModuleContext from the local environment.
+// GetConfig reads a configuration value for the module.
 //
-// This is useful for testing and development, where the environment is used to provide
-// configurations, secrets and DSNs. The context is built from a combination of
-// the ftl-project.toml file and (for now) environment variables.
-func FromEnvironment(ctx context.Context, module string) (*ModuleContext, error) {
-	cm, err := cf.NewDefaultConfigurationManagerFromEnvironment(ctx)
+// "value" must be a pointer to a Go type that can be unmarshalled from JSON.
+func (m *ModuleContext) GetConfig(name string, value any) error {
+	data, ok := m.configs[name]
+	if !ok {
+		return fmt.Errorf("no config value for %q", name)
+	}
+	return json.Unmarshal(data, value)
+}
+
+// SetConfig sets a configuration value for the module.
+func (m *ModuleContext) SetConfig(name string, value any) error {
+	data, err := json.Marshal(value)
 	if err != nil {
-		return nil, err
+		return err
 	}
-	sm, err := cf.NewDefaultSecretsManagerFromEnvironment(ctx)
+	m.SetConfigData(name, data)
+	return nil
+}
+
+// SetConfigData sets a configuration value with raw bytes
+func (m *ModuleContext) SetConfigData(name string, data []byte) {
+	m.configs[name] = data
+}
+
+// GetSecret reads a secret value for the module.
+//
+// "value" must be a pointer to a Go type that can be unmarshalled from JSON.
+func (m *ModuleContext) GetSecret(name string, value any) error {
+	data, ok := m.secrets[name]
+	if !ok {
+		return fmt.Errorf("no secret value for %q", name)
+	}
+	return json.Unmarshal(data, value)
+}
+
+// SetSecret sets a secret value for the module.
+func (m *ModuleContext) SetSecret(name string, value any) error {
```

Comment:
Do we nede setters? Can we not make the ModuleContext immutable?

If we don't, it will need locks because it could be accessed from multiple threads, but I'd vastly prefer it to be immutable.

Reply:
Will make another issue for that